### PR TITLE
make openalias also available for solo miner

### PR DESCRIPTION
**Pull request by kenshi84**
Functions listed below are moved to cryptonote_basic_impl so that monerod can also use OpenAlias:
- `simple_wallet::get_address_from_str`
- `wallet2::addresses_from_url`
- `wallet2::address_from_txt_record`
Now you can easily do solo donation mining by `start_mining getmonero.org` :D

### CODE REVIEW

**Comment by NanoAkron** [comment]
I like it

**Comment by moneromooo-monero** [comment]
I don't think this belongs here. I'd add a utiity function in, say, src/common that tries both address serialization and Openalias. Also, no std::cout in the lib please. This is going to be used by things that don't have an stdout (and also by things internally that don't want any output).

**Comment by kenshi84** [comment]
@moneromooo-monero I agree with you. I moved the functions from cryptonote_basic_impl to dns_utils. Does it look better now?

File: src/common/dns_utils.cpp [hunk 1]
```diff
@@ -323,4 +324,106 @@ bool DNSResolver::check_address_syntax(const char *addr) const
...
+namespace detail
+{
+  //-----------------------------------------------------------------------
+  // TODO: parse the string in a less stupid way, probably with regex
+  std::string address_from_txt_record(const std::string& s)
+  {
+    // make sure the txt record has "oa1:xmr" and find it
+    auto pos = s.find("oa1:xmr");
+    // search from there to find "recipient_address="
+    pos = s.find("recipient_address=", pos);
```
Comment (moneromooo-monero) [inline src/common/dns_utils.cpp]: Needs to check this doesn't return npos

File: src/common/dns_utils.cpp [hunk 2]
```diff
@@ -323,4 +324,106 @@ bool DNSResolver::check_address_syntax(const char *addr) const
...
+{
+  if (cryptonote::get_account_integrated_address_from_str(address, has_payment_id, payment_id, testnet, str_or_url))
+    return true;
+  bool dnssec_valid;
+  std::string address_str = get_account_address_as_str_from_url(str_or_url, dnssec_valid);
+  return !address_str.empty() &&
+          cryptonote::get_account_integrated_address_from_str(address, has_payment_id, payment_id, testnet, address_str);
+}
+
 }  // namespace tools
```
Comment (moneromooo-monero) [inline src/common/dns_utils.cpp]: I don't think this belongs here, but I will move stuff around myself later, don't want to spam you too much :)

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -1984,75 +1985,6 @@ bool simple_wallet::rescan_spent(const std::vector<std::string> &args)
...
-          {
-            dnssec_str = tr("WARNING: DNSSEC validation was unsuccessful, this address may not be correct!");
-          }
-          std::stringstream prompt;
-          prompt << tr("For URL: ") << url
-                 << ", " << dnssec_str << std::endl
-                 << tr(" Monero Address = ") << addresses_from_dns[0]
-                 << std::endl
-                 << tr("Is this OK? (Y/n) ")
-          ;
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Why is this removed ? Giving the user the opportunity to cancel a lookup where DNSSEC is not verified is a good thing.

File: src/common/dns_utils.cpp (hunk 1 above)
Comment (kenshi84) [inline src/common/dns_utils.cpp]: I simply copied the entire function from wallet2.cpp... I'll just add `if (pos == string::npos) return {};` then.

File: src/common/dns_utils.cpp (hunk 2 above)
Comment (kenshi84) [inline src/common/dns_utils.cpp]: I appreciate if you could take care of organizing the code. Your PR #1468 seems to have a lot of overlap with this PR, anyway.

File: src/simplewallet/simplewallet.cpp (hunk 3 above)
Comment (kenshi84) [inline src/simplewallet/simplewallet.cpp]: Because you suggested above not to use `std::cout` and `command_line::input_line` uses `std::cout`. I'm not knowledgeable at all about how the code should be organized, and I'm quite confused how I should go about this. Maybe I'm not the right person to work on this matter and this PR should be closed, or should I keep working on it?

File: src/common/dns_utils.cpp [hunk 4]
```diff
@@ -323,4 +326,140 @@ bool DNSResolver::check_address_syntax(const char *addr) const
...
+  // search from there to find "recipient_address="
+  pos = s.find("recipient_address=", pos);
+  if (pos == std::string::npos)
+    return {};
+  pos += 18; // move past "recipient_address="
+  // find the next semicolon
+  auto pos2 = s.find(";", pos);
+  if (pos2 != std::string::npos)
+  {
+    // length of address == 95, we can at least validate that much here
```
Comment (fluffypony) [inline src/common/dns_utils.cpp]: Integrated addresses are longer than 95 chars

Comment (kenshi84) [inline src/common/dns_utils.cpp]: Also accounted for the integrated address which is 106 chars long.

**Comment by moneromooo-monero** [comment]
OK with this, I'll just rejig things later.
